include/envoy/tcp/conn_pool.h

```diff
@@ -57,6 +57,18 @@ class UpstreamCallbacks : public Network::ConnectionCallbacks {
   virtual void onUpstreamData(Buffer::Instance& data, bool end_stream) PURE;
 };
 
+/**
+ * ProtocolState is a base class for protocol-specific state that must be maintained across
```

Comment:
nit: This comment is a little confusing in that it talks about state maintained across connections, but then it being destroyed when the connection is closed. Presumably it's up to the user to reattach the state to subsequent connections? Can you clarify? (Though given that we pass the pointer via move below and it's not shared I'm a little confused.)

Reply:
Sorry, it should "protocol-specific connection state that must be maintained across requests." So state per connection available across requests. I'll make it clearer with an example.

My thought was to have the connection pool caller assign a ProtocolState with a `std::unique_ptr<T>` to make clear that the connection pool owns the data, and allow it to be accessed/modified via a T* later (which might be null if there is no state yet).

Reply:
Yup makes sense.

Reply:
And actually, ConnectionState is probably a better name for this anyway. It need not actually be related to a protocol.